packages/browser/src/login/oidc/OidcLoginHandler.ts

```diff
@@ -44,6 +44,68 @@ import {
 
 import { IClient } from "@inrupt/oidc-client-ext";
 
+function isValidUrl(url: string): boolean {
+  try {
+    // Here, the URL constructor is just called to parse the given string and
+    // verify if it is a well-formed IRI.
+    // eslint-disable-next-line no-new
+    new URL(url);
+    return true;
+  } catch {
+    return false;
+  }
+}
+
+async function handleRegistration(
+  options: ILoginOptions,
+  issuerConfig: IIssuerConfig,
+  storageUtility: IStorageUtility,
+  clientRegistrar: IClientRegistrar
+): Promise<IClient> {
+  if (
+    options.clientId === undefined ||
+    (issuerConfig.solidOidcSupported === false && isValidUrl(options.clientId))
+  ) {
+    // If no client_id is provided, the client must go through DCR. Whether the
+    // Identity Provider supports Solid-OIDC or not will only change the value
+    // of the provided client_id.
+    // The tricky case is if a client WebID (here, a client_id looking like an IRI)
+    // is provided, but the Identity Provider does not support Solid-OIDC. In this
+    // case, we must discard the provided client_id, and still go through dynamic
+    // client registration, because the Identity Provider will not recognize the
+    // provided client_id.
+    return clientRegistrar.getClient(
+      {
+        sessionId: options.sessionId,
+        clientName: options.clientName,
+        redirectUrl: options.redirectUrl,
+      },
+      issuerConfig
+    );
+  }
+  // If the Identity Provider is Solid-OIDC compliant, or if the client_id isn't
+  // an IRI and therefore has been previously registered to the IdP, the client
+  // registration information just need to be stored to be retrieved after redirect.
```

Comment:
```suggestion
  // If a client_id was provided, and the Identity Provider is Solid-OIDC compliant (or it's not
  // compliant but the client_id isn't an IRI (meaning this client probably already registered with
  // the IdP via DCR)), then the client registration information needs to be stored so that it can
  // be retrieved later after redirect.
```

Reply:
0f4c45b5d22c2d0fa8f3d8ccc9055cce155fae14